tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/StreamingBlobStoreBackedReadHandleImpl.java

```diff
@@ -0,0 +1,284 @@
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.domain.Blob;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamingBlobStoreBackedReadHandleImpl implements ReadHandle {
+    private static final Logger log = LoggerFactory.getLogger(StreamingBlobStoreBackedReadHandleImpl.class);
+
+    private final long ledgerId;
+    private final List<StreamingOffloadIndexBlock> indices;
+    private final List<BackedInputStream> inputStreams;
+    private final List<DataInputStream> dataStreams;
+    private final ExecutorService executor;
+
+    static class GroupedReader {
+        long ledgerId;
+        long firstEntry;
+        long lastEntry;
+        StreamingOffloadIndexBlock index;
+        BackedInputStream inputStream;
+        DataInputStream dataStream;
+
+        public GroupedReader(long ledgerId, long firstEntry, long lastEntry,
+                             StreamingOffloadIndexBlock index,
+                             BackedInputStream inputStream, DataInputStream dataStream) {
+            this.ledgerId = ledgerId;
+            this.firstEntry = firstEntry;
+            this.lastEntry = lastEntry;
+            this.index = index;
+            this.inputStream = inputStream;
+            this.dataStream = dataStream;
+        }
+    }
+
+    private StreamingBlobStoreBackedReadHandleImpl(long ledgerId, List<StreamingOffloadIndexBlock> indices,
+                                                   List<BackedInputStream> inputStreams,
+                                                   ExecutorService executor) {
+        this.ledgerId = ledgerId;
+        this.indices = indices;
+        this.inputStreams = inputStreams;
+        this.dataStreams = new LinkedList<>();
+        for (BackedInputStream inputStream : inputStreams) {
+            dataStreams.add(new DataInputStream(inputStream));
+        }
+        this.executor = executor;
+    }
+
+    @Override
+    public long getId() {
+        return ledgerId;
+    }
+
+    @Override
+    public LedgerMetadata getLedgerMetadata() {
+        return indices.get(0).getLedgerMetadata(ledgerId);
+    }
+
+    @Override
+    public CompletableFuture<Void> closeAsync() {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        executor.submit(() -> {
+            try {
+                for (StreamingOffloadIndexBlock indexBlock : indices) {
+                    indexBlock.close();
+                }
+                for (BackedInputStream inputStream : inputStreams) {
+                    inputStream.close();
+                }
+                promise.complete(null);
+            } catch (IOException t) {
+                promise.completeExceptionally(t);
+            }
+        });
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<LedgerEntries> readAsync(long firstEntry, long lastEntry) {
+        log.debug("Ledger {}: reading {} - {}", getId(), firstEntry, lastEntry);
```

Comment:
@sijie I don't think put `log.debug` into `if (log.isDebugEnabled()) { ... }` is a good practice today. It was originally designed to reduce string construction cost in log4j1, but today every log framework implemented slf4j provided a lazy formatter to prevent string construction when debug is not enabled. It can not only reduce boilerplate code and avoid check the log level twice.

`if (log.isDebugEnabled()) { ... }` should only happened when we have extra computation for debug.

See
https://logging.apache.org/log4j/2.x/manual/api.html (Section: Substituting Parameters)
https://logging.apache.org/log4j/log4j-2.3/performance.html

Reply:
@Renkai that would be a very long discussion, thanks for bringing it up!

in my opinion it is better to have a consistent way of using the logger thru all of the codebase.

if we want to change this pattern please start a new discussion, and so we can start a new set of patches.